AX: [Site Isolation] support hit testing

hoffmanjoshua opened the pull request:
#### 3f3fcdbd0b96c559b4de3a53c56aefe356c09653
<pre>
AX: [Site Isolation] support hit testing
<a href="https://bugs.webkit.org/show_bug.cgi?id=284144">https://bugs.webkit.org/show_bug.cgi?id=284144</a>
<a href="https://rdar.apple.com/141022733">rdar://141022733</a>

Reviewed by Tyler Wilcock.

Two updates had to be made to hit testing to support site isolation/remote frames:
(1) Offset the hit test point by the remote frame offset.
(2) Return the remote frame, when applicable, from elementAccessibilityHitTest.

To validate hit testing works as expected, new AccessibilityUIElement APIs were added.
The first, `elementAtPointWithRemoteElementForTesting`, returns an AXRemoteElement
instead of a platform element, so that layout tests can check if something is a remote
element. The other, `elementAtPointResolvingRemoteFrameForTesting`, simulates how ATs
will resolve remote elements, by sending IPC messages to the remote frame via the UI
process. This method returns a debug string, to make validation easier.

Two new tests were added to verify hit testing behavior, using these new methods.

* LayoutTests/http/tests/site-isolation/accessibility/hit-test-resolving-remote-frame-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/accessibility/hit-test-resolving-remote-frame.html: Added.
* LayoutTests/http/tests/site-isolation/accessibility/hit-test-returns-remote-frame-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/accessibility/hit-test-returns-remote-frame.html: Added.
* LayoutTests/http/tests/site-isolation/resources/iframe.html: Added.
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::infoStringForTesting const):
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXRemoteFrame.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::dbgInternal const):
(WebCore::AccessibilityObject::elementAccessibilityHitTest const):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::accessibilityHitTest const):
* Source/WebCore/accessibility/AccessibilityScrollView.cpp:
(WebCore::AccessibilityScrollView::addRemoteFrameChild):
* Source/WebCore/accessibility/AccessibilityScrollView.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(attributeValueForTesting):
(-[WebAccessibilityObjectWrapper accessibilityHitTest:]):
(-[WebAccessibilityObjectWrapper _accessibilityHitTest:returnPlatformElements:]):
(-[WebAccessibilityObjectWrapper accessibilityHitTestResolvingRemoteFrame:callback:]):
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::resolveAccessibilityHitTestForTesting):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::resolveAccessibilityHitTestForTesting):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::resolveAccessibilityHitTestForTesting):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::resolveAccessibilityHitTestForTesting):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.h:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/AccessibilityUIElement.idl:
* Tools/WebKitTestRunner/InjectedBundle/ios/AccessibilityUIElementIOS.mm:
(WTR::AccessibilityUIElement::elementAtPointWithRemoteElementForTesting):
(WTR::elementAtPointResolvingRemoteFrameForTesting):
(WTR::AccessibilityUIElement::isRemoteFrame const):
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityUIElementMac.mm:
(WTR::attributeValue):
(WTR::AccessibilityUIElement::elementAtPointWithRemoteElementForTesting):
(WTR::AccessibilityUIElement::elementAtPointResolvingRemoteFrameForTesting):
(WTR::AccessibilityUIElement::isRemoteFrame const):

Canonical link: <a href="https://commits.webkit.org/287474@main">https://commits.webkit.org/287474@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/47ece6607ae0c94241dd6de68316fb53d50398b9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79791 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58792 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33190 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84316 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30797 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81900 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67866 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7083 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62366 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20209 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82859 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52424 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72667 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42707 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49766 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26815 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29241 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70894 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27286 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85742 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7025 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4921 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70644 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7199 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68506 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69865 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17415 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13874 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12792 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6983 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/12569 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6846 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10355 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8653 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -2006,8 +2006,12 @@ AccessibilityObject* AccessibilityRenderObject::accessibilityHitTest(const IntPo
         return nullptr;
 
+    // Adjust point for the remoteFrameOffset
+    IntPoint& adjustedPoint = const_cast<IntPoint&>(point);
```
This will modify the IntPoint in the scope of the caller.  Can we instead just copy point to a new IntPoint?

achristensen07 commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -3150,6 +3154,16 @@ AccessibilityObject* AccessibilityObject::elementAccessibilityHitTest(const IntP
         }
+
+        if (widget && widget->isRemoteFrameView()) {
+            if (AXObjectCache* cache = axObjectCache()) {
```
CheckedPtr

achristensen07 commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -3150,6 +3154,16 @@ AccessibilityObject* AccessibilityObject::elementAccessibilityHitTest(const IntP
+
+        if (widget && widget->isRemoteFrameView()) {
+            if (AXObjectCache* cache = axObjectCache()) {
+                if (auto* remoteHostWidget = cache->getOrCreate(*widget)) {
```
RefPtr

achristensen07 commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -3150,6 +3154,16 @@ AccessibilityObject* AccessibilityObject::elementAccessibilityHitTest(const IntP
+            if (AXObjectCache* cache = axObjectCache()) {
+                if (auto* remoteHostWidget = cache->getOrCreate(*widget)) {
+                    remoteHostWidget->updateChildrenIfNecessary();
+                    auto* scrollView = dynamicDowncast<AccessibilityScrollView>(remoteHostWidget);
```
RefPtr

achristensen07 commented on Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
```diff
@@ -2464,6 +2476,44 @@ - (id)accessibilityHitTest:(NSPoint)point
+- (void)accessibilityHitTestResolvingRemoteFrame:(NSPoint)point callback:(void(^)(NSString *))callback
+{
+    if (!AXObjectCache::clientIsInTestMode())
+        return;
```
Do we want all flow paths to call callback?

hoffmanjoshua replied:
Yeah, it probably can't hurt. I mostly have this to prevent non-testing clients from using `accessibilityHitTestResolvingRemoteFrame`, so we should just give those empty strings as parameters.

achristensen07 commented on Source/WebCore/page/ChromeClient.h:
```diff
@@ -563,6 +563,8 @@ class ChromeClient {
 #if ENABLE(ACCESSIBILITY_ANIMATION_CONTROL)
     virtual void isAnyAnimationAllowedToPlayDidChange(bool /* anyAnimationCanPlay */) { };
 #endif
+    virtual void resolveAccessibilityHitTestForTesting(WebCore::FrameIdentifier, const IntPoint&, CompletionHandler<void(String)>&&) { };
```
Can we make this pure virtual or call the CompletionHandler with an empty string in the empty client case?

hoffmanjoshua replied:
Yes that's a good call

achristensen07 commented on Tools/WebKitTestRunner/InjectedBundle/ios/AccessibilityUIElementIOS.mm:
```diff
@@ -258,6 +258,17 @@ - (CGPathRef)_accessibilityPath;
     return AccessibilityUIElement::create(element);
 }
 
+
```
whitespace

twilco commented on LayoutTests/http/tests/site-isolation/accessibility/hit-test-resolving-remote-frame.html:
```diff
@@ -0,0 +1,38 @@
+        window.jsTestIsAsync = true;
+
+        const iframe = document.getElementById("iframe");
+        const iframeRect = iframe.getBoundingClientRect();
```
Since we only use `iframe` to get its rect, I think shortening these two lines to this is a bit simpler:

`const iframeRect = document.getElementById("iframe").getBoundingClientRect();`

twilco commented on LayoutTests/http/tests/site-isolation/accessibility/hit-test-resolving-remote-frame.html:
```diff
@@ -0,0 +1,38 @@
+        const iframe = document.getElementById("iframe");
+        const iframeRect = iframe.getBoundingClientRect();
+
+        output += `iFrame rect: ${iframeRect.x}, ${iframeRect.y}\n`;
```
I think iframes are typically not camel-cased, e.g. do we want `iframe rect:` instead of `iFrame rect:`?

hoffmanjoshua replied:
Yeah, I updated the other test but missed this one—will fix this

twilco commented on LayoutTests/http/tests/site-isolation/accessibility/hit-test-resolving-remote-frame.html:
```diff
@@ -0,0 +1,38 @@
+
+        output += `iFrame rect: ${iframeRect.x}, ${iframeRect.y}\n`;
+
+        var text;
```
Is this used?

hoffmanjoshua replied:
Nope, I can remove this.

hoffmanjoshua replied:
We also don't need to declare `axElement` below this, now that the test structure has changed. I'll update that as well.

twilco commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -3141,15 +3145,25 @@ bool AccessibilityObject::supportsARIAAttributes() const
+            if (CheckedPtr cache = axObjectCache()) {
+                if (RefPtr remoteHostWidget = cache->getOrCreate(*widget)) {
+                    remoteHostWidget->updateChildrenIfNecessary();
+                    RefPtr scrollView = dynamicDowncast<AccessibilityScrollView>(remoteHostWidget);
```
Adding a `*` to this cast will save a null-check, i.e. `dynamicDowncast<...>(*remoteHostWidget)`

twilco commented on Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.h:
```diff
@@ -137,6 +137,8 @@ extern "C" AXUIElementRef NSAccessibilityCreateAXUIElementRef(id element);
 // When a plugin uses a WebKit control to act as a surrogate view (e.g. PDF use WebKit to create text fields).
 - (id)_associatedPluginParent;
+// For testing use only.
+- (void)accessibilityHitTestResolvingRemoteFrame:(NSPoint)point callback:(void(^)(NSString *))callback;
```
We should prefix this with `_` to indicate it's "private".

hoffmanjoshua replied:
Good call!

twilco commented on Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
```diff
@@ -2464,6 +2476,44 @@ - (id)accessibilityHitTest:(NSPoint)point
+
+            auto clientCallback = [callback = makeBlockPtr(callback)] (String result) {
+                callback(nsStringNilIfEmpty(result));
+            };
```
Doing this is unnecessary if `!page` or `!axRemoteFrame` — can we move it inside that scope?

```cpp
if (page && axRemoteFrame) {
    auto clientCallback = [callback = makeBlockPtr(callback)] (String result) {
        callback(nsStringNilIfEmpty(result));
    };
    page->chrome().client().resolveAccessibilityHitTestForTesting(*axRemoteFrame->frameID(), IntPoint(point), WTFMove(clientCallback));
}
```

twilco commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -474,6 +474,21 @@
+        }
+    }
+#endif
+    completionHandler(makeString("NULL"_s));
```
I'm surprised there are no compiler warnings about `point` being unused on `!PLATFORM(MAC)` 🤔 Would've thought we needed an `UNUSED_PARAM(point)`

hoffmanjoshua replied:
There probably will be 😅 , I've just been fixing other errors on `!PLATFORM(MAC)` so it has been masked. So I'll add the unused param in.

twilco commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -1639,6 +1639,10 @@ void WebPage::bindRemoteAccessibilityFrames(int, WebCore::FrameIdentifier, Vecto
 {
 }
 
+void WebPage::resolveAccessibilityHitTestForTesting(const WebCore::IntPoint& point, CompletionHandler<void(String)>&& completionHandler)
```
I think when we don't use function arguments at all, we don't give them a name. So:

```cpp
void WebPage::resolveAccessibilityHitTestForTesting(const WebCore::IntPoint&, CompletionHandler<void(String)>&&)
{
}
```

twilco commented on Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityUIElementMac.mm:
```diff
@@ -542,6 +546,34 @@ void setAttributeValue(id element, NSString* attribute, id value, bool synchrono
+        element = [m_element _accessibilityHitTest:NSMakePoint(x, y) returnPlatformElements:NO];
+    });
+
+    if (!element)
+        return nullptr;
+
+    return AccessibilityUIElement::create(element.get());
```
Could be a bit shorter as:

`return element ? AccessibilityUIElement::create(element.get()) : nullptr;`

hoffmanjoshua replied:
Unfortunately the compiler doesn't agree with this one, so going to have to leave it as-is for now.

hoffmanjoshua committed:
AX: [Site Isolation] support hit testing
https://bugs.webkit.org/show_bug.cgi?id=284144
rdar://141022733

Reviewed by Tyler Wilcock.

Two updates had to be made to hit testing to support site isolation/remote frames:
(1) Offset the hit test point by the remote frame offset.
(2) Return the remote frame, when applicable, from elementAccessibilityHitTest.

To validate hit testing works as expected, new AccessibilityUIElement APIs were added.
The first, `elementAtPointWithRemoteElementForTesting`, returns an AXRemoteElement
instead of a platform element, so that layout tests can check if something is a remote
element. The other, `elementAtPointResolvingRemoteFrameForTesting`, simulates how ATs
will resolve remote elements, by sending IPC messages to the remote frame via the UI
process. This method returns a debug string, to make validation easier.

Two new tests were added to verify hit testing behavior, using these new methods.

* LayoutTests/http/tests/site-isolation/accessibility/hit-test-resolving-remote-frame-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/accessibility/hit-test-resolving-remote-frame.html: Added.
* LayoutTests/http/tests/site-isolation/accessibility/hit-test-returns-remote-frame-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/accessibility/hit-test-returns-remote-frame.html: Added.
* LayoutTests/http/tests/site-isolation/resources/iframe.html: Added.
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::infoStringForTesting const):
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXRemoteFrame.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::dbgInternal const):
(WebCore::AccessibilityObject::elementAccessibilityHitTest const):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::accessibilityHitTest const):
* Source/WebCore/accessibility/AccessibilityScrollView.cpp:
(WebCore::AccessibilityScrollView::addRemoteFrameChild):
* Source/WebCore/accessibility/AccessibilityScrollView.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(attributeValueForTesting):
(-[WebAccessibilityObjectWrapper accessibilityHitTest:]):
(-[WebAccessibilityObjectWrapper _accessibilityHitTest:returnPlatformElements:]):
(-[WebAccessibilityObjectWrapper accessibilityHitTestResolvingRemoteFrame:callback:]):
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::resolveAccessibilityHitTestForTesting):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::resolveAccessibilityHitTestForTesting):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::resolveAccessibilityHitTestForTesting):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::resolveAccessibilityHitTestForTesting):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.h:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/AccessibilityUIElement.idl:
* Tools/WebKitTestRunner/InjectedBundle/ios/AccessibilityUIElementIOS.mm:
(WTR::AccessibilityUIElement::elementAtPointWithRemoteElementForTesting):
(WTR::elementAtPointResolvingRemoteFrameForTesting):
(WTR::AccessibilityUIElement::isRemoteFrame const):
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityUIElementMac.mm:
(WTR::attributeValue):
(WTR::AccessibilityUIElement::elementAtPointWithRemoteElementForTesting):
(WTR::AccessibilityUIElement::elementAtPointResolvingRemoteFrameForTesting):
(WTR::AccessibilityUIElement::isRemoteFrame const):

Canonical link: https://commits.webkit.org/287474@main
